depext.ml

```diff
@@ -67,182 +75,198 @@ let ask ?(default=false) fmt =
-     | _           -> `Unix)
-  | "Win32"  -> `Win32
-  | "Cygwin" -> `Cygwin
-  | s        -> `Other s
-
-let distribution = function
-  | `Darwin ->
-    if has_command "brew" then Some `Homebrew
-    else if has_command "port" then Some `Macports
-    else None
-  | `Linux ->
+let has_prefix s pfx =
+  let pfxlen = String.length pfx in
+  pfxlen <= String.length s &&
+  try for i = 0 to pfxlen do
+      if pfx.[i] <> s.[i] then raise Exit
+    done;
+    true
+  with Exit -> false
+
+let arch =
+  let raw = match Sys.os_type with
+    | "Unix" | "Cygwin" -> command_output "uname -m"
+    | "Win32" ->
+      (match Sys.getenv "PROCESSOR_ARCHITECTURE" with
+       | "X86" as a ->
+         (try Sys.getenv "PROCESSOR_ARCHITEW6432" with Not_found -> a)
+       | arch -> arch)
+    | _ -> failwith "Bad Sys.os_type"
+  in
+  match String.lowercase raw with
+  | "x86" | "i386" | "i586" | "i686" -> "x86_32"
+  | "x86_64" | "amd64" -> "x86_64"
+  | "powerpc" | "ppc" | "ppcle" -> "ppc32"
+  | "ppc64" | "ppc64le" -> "ppc64"
+  | "aarch64_be" | "aarch64" | "armv8b" | "armv8l" -> "arm64"
+  | a when List.exists (has_prefix a)
+        ["armv5"; "armv6"; "earmv6"; "armv7"; "earmv7"] -> "arm32"
+  | "" -> "unknown"
+  | s -> s
+
+let os =
+  match Sys.os_type with
+  | "Unix" -> (match String.lowercase (command_output "uname -s") with
+      | "darwin" -> "macos"
+      | "" -> "unknown"
+      | s -> s)
+  | s -> String.lowercase s
+
+let os_release_field: string -> string =
+  let os_release_file = lazy (
+    List.find Sys.file_exists ["/etc/os-release"; "/usr/lib/os-release"] |>
+    lines_of_file |>
+    List.map (fun s -> Scanf.sscanf s "%s@= %s" (fun x v ->
+        x,
+        try Scanf.sscanf v "\"%s@\"" (fun s -> s)
+        with Scanf.Scan_failure _ -> v))
+  ) in
+  fun f ->
+    List.assoc f (Lazy.force os_release_file)
+
+let is_android, android_release =
+  let prop = lazy (
+    command_output "getprop ro.build.version.release 2>/dev/null"
+  ) in
+  (fun () -> try ignore (Lazy.force prop); true with Failure _ -> false),
+  (fun () -> Lazy.force prop)
+
+let distribution =
+  match os with
+  | "macos" ->
+    if has_command "brew" then "homebrew"
+    else if has_command "port" then "macports"
+    else os
+  | "linux" ->
+    (String.lowercase @@
+     if is_android () then "android" else
+     try os_release_field "ID" with Not_found ->
+     try command_output "lsb_release -i -s 2>/dev/null" with Failure _ ->
+     try
+       List.find Sys.file_exists ["/etc/redhat-release";
+                                  "/etc/centos-release";
+                                  "/etc/gentoo-release";
+                                  "/etc/issue"] |>
+       fun s -> Scanf.sscanf s " %s " (fun s -> s)
+     with Not_found -> os)
+  | _ -> os
+
+let os_version =
+  match os with
+  | "linux" ->
+    (String.lowercase @@
+     try android_release () with Failure _ ->
+     try command_output "lsb_release -s -r" with Failure _ ->
```

Comment:
This lsb_release needs to redirect stderr, or else you get `lsb_release: command not found` on (e.g.) Alpine